### ISSUE TITLE
Logitech Flow

### ISSUE THREAD

**Issue by henioStraszny** [pwr-Solaar/Solaar#672]
**Information**
- Output of `solaar show` for the target device (if applicable):
```
 solaar show                                                                                                                                            
Unifying Receiver                                                                                                                                               
  Device path  : /dev/hidraw5                                                                                                                                   
  USB id       : 046d:c52b                                                                                                                                      
  Serial       : 2DC1908E                                                                                                                                       
    Firmware   : 12.10.B0032                                                                                                                                    
    Bootloader : 02.14                                                                                                                                          
    Other      : AA.AA                                                                                                                                          
  Has 3 paired device(s) out of a maximum of 6.                                                                                                                 
  Notifications: wireless, software present (0x000900)                                                                                                          
  Device activity counters: 2=83                                                
                                                                                
  1: Wireless Touch Keyboard K400                                               
     Codename     : K400                                                        
     Kind         : keyboard                                                    
     Wireless PID : 4024      
     Protocol     : HID++ 2.0  
     Polling rate : 8 ms (125Hz)                                                                                                                                
     Serial number: 5A0D1FFA                                                    
     The power switch is located on the edge of top right corner.                                                                                               
     Battery: unknown (device is offline).               
                                                                                                                                                                
  2: M585/M590 Multi-Device Mouse                                                                                                                               
     Codename     : M585/M590                                                                                                                                   
     Kind         : mouse                                                                                                                                       
     Wireless PID : 406B                                                                                                                                        
     Protocol     : HID++ 4.5                                                                                                                                   
     Polling rate : 8 ms (125Hz)                                                                                                                                
     Serial number: 1CD931FC                                                                                                                                    
        Bootloader: BOT 48.00.B0001                                                                                                                             
          Firmware: MPM 05.01.B0006                                                                                                                             
             Other:                                                                                                                                             
     The power switch is located on the base.            
     Supports 33 HID++ 2.0 features:
         0: ROOT                   {0000}   
         1: FEATURE SET            {0001}   
         2: DEVICE FW VERSION      {0003}   
         3: DEVICE NAME            {0005}   
         4: WIRELESS DEVICE STATUS {1D4B}   
         5: RESET                  {0020}   
         6: unknown:0007           {0007}   
         7: BATTERY STATUS         {1000}   
         8: CHANGE HOST            {1814}    
         9: unknown:1815           {1815}   
        10: REPROG CONTROLS V4     {1B04}   
        11: unknown:1C00           {1C00}                                                                                                                       
        12: POINTER SPEED          {2205}   
            Pointer Speed: 1.0                                                                                                                                  
        13: VERTICAL SCROLLING     {2100}   
            Roller type: standard                                                                                                                               
            Ratchet per turn: 18 
            Scroll lines: 0                                                                                                                                     
        14: unknown:00C2           {00C2}   
        15: unknown:1802           {1802}   internal, hidden                                                                                                    
        16: unknown:1803           {1803}   internal, hidden
        17: unknown:1806           {1806}   internal, hidden   
 18: unknown:1805           {1805}   internal, hidden                                                                                                    
        19: unknown:1813           {1813}   internal, hidden                                                                                                    
        20: unknown:1830           {1830}   internal, hidden                                                                                                    
        21: unknown:1861           {1861}   internal, hidden                                                                                                    
        22: unknown:1890           {1890}   internal, hidden                                                                                                    
        23: unknown:1891           {1891}   internal, hidden                                                                                                    
        24: unknown:18A1           {18A1}   internal, hidden                                                                                                    
        25: unknown:1DF3           {1DF3}   internal, hidden                                                                                                    
        26: unknown:1E00           {1E00}   hidden                                                                                                              
        27: unknown:1EB0           {1EB0}   internal, hidden                                                                                                    
        28: unknown:18B1           {18B1}   internal, hidden                    
        29: unknown:1850           {1850}   internal, hidden                    
        30: unknown:1F03           {1F03}   internal, hidden                    
        31: unknown:18C0           {18C0}   internal, hidden                    
        32: LOWRES WHEEL           {2130}                                       
            Wheel Reports: HID
     Has 8 reprogrammable keys:
         0: LEFT CLICK                , default: LeftClick                   => LEFT CLICK                                                                      
             mse, reprogrammable, pos:0, group:1, gmask:1                       
         1: RIGHT CLICK               , default: RightClick                  => RIGHT CLICK                                                                     
             mse, reprogrammable, pos:0, group:1, gmask:1
         2: MIDDLE BUTTON             , default: MiddleMouseButton           => MIDDLE BUTTON                                                                   
             divertable, mse, persistently divertable, reprogrammable, pos:0, group:2, gmask:3                                                                  
         3: BACK AS BUTTON 4          , default: BackEx                      => BACK AS BUTTON 4                                                                
             divertable, mse, persistently divertable, reprogrammable, pos:0, group:2, gmask:3                                                                  
         4: FORWARD AS BUTTON 5       , default: BrowserForwardEx            => FORWARD AS BUTTON 5                                                             
             divertable, mse, persistently divertable, reprogrammable, pos:0, group:2, gmask:3                                                                  
         5: LEFT SCROLL AS AC PAN     , default: HorzScrollLeftSet           => LEFT SCROLL AS AC PAN                                                           
             divertable, mse, persistently divertable, reprogrammable, pos:0, group:2, gmask:3                                                                  
         6: RIGHT SCROLL AS AC PAN    , default: HorzScrollRightSet          => RIGHT SCROLL AS AC PAN                                                          
             divertable, mse, persistently divertable, reprogrammable, pos:0, group:2, gmask:3                                                                  
         7: unknown:00D7              , default: unknown:00B4                => unknown:00D7                                                                    
             divertable, virtual, pos:0, group:3, gmask:0
     Battery: 50%, discharging.     
                                                                                
  3: Wireless Mouse MX Master 2S
     Codename     : MX Master 2S
     Kind         : mouse
     Wireless PID : 4069
     Protocol     : HID++ 4.5
     Polling rate : 8 ms (125Hz)
     Serial number: 012EDC4E
        Bootloader: BOT 56.01.B0006
          Firmware: MPM 12.01.B0006
          Firmware: MPM 12.01.B0006
             Other: 
     The power switch is located on the base.
     Supports 32 HID++ 2.0 features:
         0: ROOT                   {0000}   
         1: FEATURE SET            {0001}   
         2: DEVICE FW VERSION      {0003}   
         3: DEVICE NAME            {0005}   
         4: WIRELESS DEVICE STATUS {1D4B}   
         5: RESET                  {0020}   
         6: unknown:0021           {0021}   
         7: BATTERY STATUS         {1000}   
         8: unknown:1806           {1806}   internal, hidden
         9: CHANGE HOST            {1814}   
        10: REPROG CONTROLS V4     {1B04}   
        11: ADJUSTABLE DPI         {2201}   
        12: VERTICAL SCROLLING     {2100}   
            Roller type: 3G
            Ratchet per turn: 24
            Scroll lines: 0
        13: SMART SHIFT            {2110}   
        14: HIRES WHEEL            {2121}   
            Multiplier: 8
            Has invert
              Normal wheel motion
            Has ratchet switch
              Normal wheel mode
            High resolution mode
            HID notification
        15: GESTURE 2              {6501}   
        16: unknown:00C2           {00C2}   
        17: unknown:1813           {1813}   internal, hidden
        18: unknown:1830           {1830}   internal, hidden
        19: unknown:1890           {1890}   internal, hidden
        20: unknown:1891           {1891}   internal, hidden
        21: unknown:18A1           {18A1}   internal, hidden
        22: unknown:18C0           {18C0}   internal, hidden
        23: unknown:1DF3           {1DF3}   internal, hidden
        24: unknown:1E00           {1E00}   hidden
        25: unknown:1EB0           {1EB0}   internal, hidden
        26: unknown:1803           {1803}   internal, hidden
        27: unknown:1861           {1861}   internal, hidden
        28: unknown:9001           {9001}   internal, hidden
        29: unknown:9200           {9200}   internal, hidden
        30: unknown:9202           {9202}   internal, hidden
        31: unknown:1805           {1805}   internal, hidden
     Has 8 reprogrammable keys:
         0: LEFT CLICK                , default: LeftClick                   => LEFT CLICK                
             mse, pos:0, group:1, gmask:1
         1: RIGHT CLICK               , default: RightClick                  => RIGHT CLICK               
             mse, pos:0, group:1, gmask:1
         2: MIDDLE BUTTON             , default: MiddleMouseButton           => MIDDLE BUTTON             
             divertable, mse, reprogrammable, pos:0, group:3, gmask:7
         3: BACK AS BUTTON 4          , default: BackEx                      => BACK AS BUTTON 4          
             divertable, mse, reprogrammable, pos:0, group:2, gmask:3
         4: FORWARD AS BUTTON 5       , default: BrowserForwardEx            => FORWARD AS BUTTON 5       
             divertable, mse, reprogrammable, pos:0, group:2, gmask:3
         5: unknown:00C3              , default: unknown:00A9                => unknown:00C3              
             divertable, mse, reprogrammable, pos:0, group:3, gmask:7
         6: unknown:00C4              , default: unknown:009D                => unknown:00C4              
             divertable, mse, reprogrammable, pos:0, group:3, gmask:7
         7: unknown:00D7              , default: unknown:00B4                => unknown:00D7              
             divertable, virtual, pos:0, group:4, gmask:0
     Battery: 90%, discharging.


```

**Is your feature request related to a problem? Please describe.**
Solaar does not support Logitech Flow.

**Describe the solution you'd like**
Logitech Flow is supported.

**Describe alternatives you've considered**
Logitech Flow consists of two parts:
1. Possibility to move cursor between two physical computers.
2. Copy Paste text, files etc.
Perhaps event point 1 would be of great benefit for users.

**Additional context**
- Logitech official feature explanation: https://www.logitech.com/en-us/product/options/page/flow-multi-device-control
- In this ticket: #367, users were asking about this functionality.
- This feature is available for MX Series mice and keyboards.

**Comment by FFY00** [comment]
That is out of scope. I don't really see a reason to have a hybrid software/hardware implementation of a virtual KVM switch when proper all software solutions exist.

Take a look at https://github.com/debauchee/barrier or https://github.com/symless/synergy-core. They should do exactly what you want.

Logitech Flow is just seems to be a marketing scheme to make buyers of high-end devices feel like they are getting more out of them. It doesn't make sense from a technical POV.

With that said, in the upcoming logitechd daemon I am more than happy to provide access to the hardware settings that Logitech Flow utilizes. Which AFAIK only allow you to change the active host in a multi-host device.

**Comment by pfps** [comment]
Closing as out of scope.

**Comment by ctp-placebo** [comment]
Logitech Flow allows copying from the clipboard on one PC and pasting on a 2nd PC. Using the USB-IR receiver on PC1, and Bluetooth on PC2, there is a button on top of the Ergo MX Pro trackball mouse to switch between the two receivers. 
Copy on PC1, switch receivers, paste on PC2. 
It's "not just a marketing tool", it is a function of the hardware that we would like to see Logitech support on Linux!

**Comment by pfps** [comment]
The cut-and-paste part of Logitech Flow requires Logitech software running on both machines, as far as I can tell.  The communications protocol that is used is unknown.  Synergy has its own way of multi-machine cut-and-paste.

**Comment by barolo** [comment]
@pfps @ctp-placebo I think that it's worth mentioning in the docs that there's already massively popular software that includes inter-system cut-and-paste that's not only multiOS [ including Android  ] but is adapted for pretty much every major desktop, namely KDE Connect.

**Comment by sperry04** [comment]
I know this issue is closed, but I will point out there is a fundamental difference between Flow and software packages like Synergy/etc. that allow keyboard/mouse sharing that might make this feature worth implementing.

- With Flow, the software only triggers the device switch on the hardware, so:
  - no network latency/traffic for client computers
  - no security risk of keystroke sniffing on the network
  - no software keyboard/mouse device emulation occurring on the client computers
  - if the server computer is off, keyboard/mouse continue to work with connected computers

It seems to me that Flow might be functionally similar to keyboard/mouse sharing apps, but has some advantages due to the hardware implementation aspect.

**Comment by barolo** [comment]
@sperry04 Flow needs network for some of its functionality, it won't turn on without it.
Solaar's implementation is on hardware's level and works without network too.

**Comment by ryandasilva** [comment]
**### You know what this sounds like to me? A group of programmers who may come off as pretentious and unwilling to tackle a unique, yet simple, and somewhat complicated feature. Some say it requires network access? So what? We have printer setup software that swiftly scans the network for devices that respond on TCP or UDP ports, making wired or Wi-Fi printer setup a breeze. This applies to both new out-of-the-box setups and printers already in use, allowing additional users to connect.
My knowledge of network programming is limited to creating a new TCP socket and performing a basic handshake. I've even developed a fully functional peer-to-peer system and was close to finishing a client-server messaging service, somewhat similar to ICQ from the past. But I lacked the understanding of the broader internet and NAT. It worked well within the same LAN, didn't handle NAT, and was fast and lightweight.**

I'm not here to boast or belittle anyone. I came here because Google directed me, indicating there's demand for this feature. It's October 2023, and more devices support Flow, while Linux usage continues to grow. I rely on Solaar, so I'm not here to criticize you. My critiques are a sign of my respect for this community and belief that the responses against implementing Flow in Solaar seem weak, given the talent I've seen in the open-source/Linux community.

If you're determined, an alpha release could be done in a long weekend, a beta by the end of the month, and a public release in less than six weeks. Correct me if I'm wrong, but doesn't Flow work with MacOS and Windows? If so, much of the work is already done; just reverse engineer the MacOS version since it's essentially BSD code.

I'd help if I could, but it's been 18 years since I last wrote amateur software, and I don't have time to learn everything needed now. The world has changed a lot since then, with GitHub and many skilled programmers worldwide.

I hope this message motivates someone to try. If not, at least I've tried in the capacity I could, unlike those who dismissed this feature request arrogantly.

Stay safe. Best regards, cheers, and all the best!
